icicle/backend/cpu/src/hash/cpu_poseidon.cpp

```diff
@@ -1,47 +1,307 @@
+
+      unsigned int alpha = poseidon_constants[T].alpha;
+      unsigned int nof_upper_full_rounds = poseidon_constants[T].nof_upper_full_rounds;
+      unsigned int nof_partial_rounds = poseidon_constants[T].nof_partial_rounds;
+      unsigned int nof_bottom_full_rounds = poseidon_constants[T].nof_bottom_full_rounds;
+      S* rounds_constants = poseidon_constants[T].rounds_constants;
+      S* mds_matrix = poseidon_constants[T].mds_matrix;
+      S* pre_matrix = poseidon_constants[T].pre_matrix;
+      S* sparse_matrices = poseidon_constants[T].sparse_matrices;
+      // Allocate temporary memory for intermediate calcs.
+      S* tmp_fields = new S[T];
+      // Casting from bytes to scalar.
+      const S* in_fields = (S*)(input);
+      // Copy input scalar to the output (as a temp storage) to be used in the rounds.
+      // *tmp_fields are used as a temp storage during the calculations in this function.
+      memcpy(tmp_fields, in_fields, T * sizeof(S));
+
+      // Add pre-round constants.
+      for (int state_idx = 0; state_idx < T; state_idx++) {
+        tmp_fields[state_idx] = tmp_fields[state_idx] + *rounds_constants++;
+      }
+
+      // Upper full rounds.
+      full_rounds(nof_upper_full_rounds, tmp_fields, rounds_constants);
+
+      // Single full round with pre_matrix.
+      for (int state_idx = 0; state_idx < T; state_idx++) {
+        // S box
+        tmp_fields[state_idx] = S::pow(tmp_fields[state_idx], alpha);
+      }
+      for (int state_idx = 0; state_idx < T; state_idx++) {
+        // Add round constants
+        tmp_fields[state_idx] = tmp_fields[state_idx] + *rounds_constants++;
+      }
+      // Multiplication by pre_matrix
+      field_vec_sqr_full_matrix_mul(tmp_fields, pre_matrix, tmp_fields);
+
+      // Partial rounds. Perform calculation only for the first element of *tmp_fields.
+      for (int partial_rounds_idx = 0; partial_rounds_idx < nof_partial_rounds; partial_rounds_idx++) {
+        // S box
+        tmp_fields[0] = S::pow(tmp_fields[0], alpha);
+        // Add round constants
+        tmp_fields[0] = tmp_fields[0] + *rounds_constants++;
+        // Multiplication by sparse matrix.
+        field_vec_sqr_sparse_matrix_mul(tmp_fields, &sparse_matrices[partial_rounds_idx * (2 * T - 1)], tmp_fields);
+      }
+
+      // Bottom full rounds.
+      full_rounds(nof_bottom_full_rounds, tmp_fields, rounds_constants);
+
+      // Last full round
+      for (int state_idx = 0; state_idx < T; state_idx++) {
+        // S box
+        tmp_fields[state_idx] = S::pow(tmp_fields[state_idx], alpha);
+      }
+      // Multiplication by MDS matrix
+      field_vec_sqr_full_matrix_mul(tmp_fields, mds_matrix, tmp_fields);
+
+      memcpy(output, (std::byte*)(&tmp_fields[1]), sizeof(S));
+
+      delete[] tmp_fields;
+      tmp_fields = nullptr;
+
       return eIcicleError::SUCCESS;
     }
-  };
+
+    void field_vec_sqr_sparse_matrix_mul(const S* vec_in, const S* matrix_in, S* result) const
```

Comment:
add comment to describe what that function does

Reply:
Done.